Move Ocr into seperate compiled daemon

GeekTeaa opened the pull request:
Moved all the OCR engine into it's own directory and is being compiled into it's own image. This will be the start of the OCR Server. In addition to this, I created modules necessary to make that process a daemon, and a log module. That image will enter the `StartServer` and exit. 

Forewent any efforts to get the server to listen on the socket and run a received image through the OCR. This PR was already large enough.

GeekTeaa committed:
Moved the src into gui folder so I can start seperating the OCR api
into it's own module.

GeekTeaa committed:
Added ocrserver dir for all ocr needs. Need to add crow now.

GeekTeaa committed:
Began development on OCR server. Properly linking cpprest library, but
need to clean that up before I merge to develop branch. Added
logger module, and signals module for handling module. Tell future me
to commit for often, this commit is too big.

GeekTeaa committed:
Refactored log file, add ocr_server, daemonized main.cpp

GeekTeaa committed:
Renmaed ocrserver directory to server directory

GeekTeaa committed:
Updated makefile to work better with other people's systems... In case
anyone feels like contributing...

GeekTeaa committed:
Added a sequence diagram, Added module that daemonizes the process

GeekTeaa committed:
Finished daemonize_process.c, doing basic signal handling though this
module will need to updated. We get to a point where we can start the
server, and that's where we'll listen and respond to messages.

GeekTeaa committed:
Delete Makefile

Makefile is generated by Qmake and should not be in the repository

matthew-rindel commented on server/ocr_server_signals.cpp:
```diff
@@ -0,0 +1,50 @@
+// e.x. SignalHandler.addSignalResponse(SIGHUP, someFunctionPointer);
+
+static signals::Signal all_signals[] = {
+  {.sig = SIGHUP, .func = HandleSegmentationFaultSignal},
```
It looks like the signal handlers may be mixed up here, between hangup and segmentation fault.

matthew-rindel commented on server/ocr_server.cpp:
```diff
@@ -0,0 +1,9 @@
+#include "lib/log.hpp"
+#include <unistd.h>
```
unistd.h may cause trouble on Windows

matthew-rindel commented on server/ocr_server.cpp:
```diff
@@ -0,0 +1,9 @@
+void ocrserver::OcrServer::StartServer(void) {
+  sys::Log log;
+  log.Info("Starting OCR Server\r\n");
+  sleep(10);
```
This is a bit longer: std::this_thread::sleep_for(std::chrono::milliseconds(10));

But is portable and is available from C++ 11 onwards. 

The thread and chrono headers would be needed.

matthew-rindel commented on server/daemonize_process.c:
```diff
@@ -0,0 +1,98 @@
+#include "daemonize_process.h"
+#include <stdio.h>
+#include <stdlib.h>
+#include <unistd.h>
```
I found this, which I will look at using for Windows: https://github.com/robinrowe/libunistd

Though the author of the library suggests using pthreads or C++ threads instead of fork. Perhaps something to consider.

matthew-rindel commented on lib/log.cpp:
```diff
@@ -0,0 +1,65 @@
+#include <stdio.h>
+#include <stdarg.h>
+#include <syslog.h>
```
I knew I wasn't familiar with this. Looks like it will also be an issue on Windows. 

No need to change the code though, as there is this: https://github.com/asankah/syslog-win32

Probably more shims like it out there. I'll keep a look out.

GeekTeaa committed:
Removed dependency on unistd from a majority of the module, Added some
TODOs, finished syslog work

GeekTeaa committed:
Merge branch 'move-ocr-into-seperate-compiled-daemon' of github.com:GeekTeaa/LangOCR into move-ocr-into-seperate-compiled-daemon

GeekTeaa committed:
Neglected to initialize signal handlers, added the call to do
so. Fixed silly mix up where the wrong signal handler was being called
for the wrong signal.
